### PR TITLE
Refactor crypto be strongly typed

### LINKED ISSUE #193: Refactor crypto be strongly typed and enforce encryption at rest
**Motivation**
For existing fields that are encrypted (user.secret and zoneConnection.key), these fields are Stringly typed.  There is no way to understand their disposition when we save / retrieve the values.

We should have compile time / type checks on the disposition of these values so they are handled properly in the code and we do not accidentally surface a value decrypted when it should not be.

**Scope of change**
1. Create a strongly typed value to represent a `String` as being encrypted or decrypted.
1. Update the `ZoneConnection` type to use the new encrypted type
1. Update the `User` type to use the new encrypted type

**Design Notes**

Some sketches from a design perspective...

```scala
// I thought of having a Decrypted here as well, but in reality we should never have the 
// data classes contain decrypted data.  This makes sure we don't accidentally surface decrypted data
// We need a way to pull the decrypted value out of the Encrypted instance though elsewhere
final case class Encrypted private (value: String) extends AnyVal

object Encrypted { 
  def apply(crypto: CryptoAlgebra, value: String): Encrypted = new Encrypted(crypto.encrypt(value))
  def decrypt(crypto: CryptoAlgebra, x: Encrypted): String = crypto.decrypt(x.value)
}
```

### PR DESCRIPTION
**Pull request by Aravindh-Raju**
Fixes #193.

Changes in this pull request:
- Created a strongly typed value to represent a `String` as being encrypted or decrypted.
- Updated the `ZoneConnection` type to use the new encrypted type.
- Updated the `User` type to use the new encrypted type.

### CODE REVIEW

File: modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala [hunk 1]
```diff
@@ -129,12 +130,22 @@ trait DnsJsonProtocol extends JsonValidation {
     override def toJson(a: Algorithm): JValue = JString(a.name)
   }
 
+  case object EncryptedSerializer extends ValidationSerializer[Encrypted] {
```
Comment (nspadaccino) [inline modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala:133]: Do we not have unit tests for DnsJsonProtocol? It seems like we do for other serializers, I think that would be a good addition

**Review by nspadaccino — COMMENTED** [review]
Looks good, tested locally via API and portal, I am able authenticate with test creds. One comment about unit tests

File: modules/api/src/it/scala/vinyldns/api/domain/record/RecordSetServiceIntegrationSpec.scala [hunk 2]
```diff
@@ -33,7 +33,7 @@ import vinyldns.api.domain.zone._
 import vinyldns.api.engine.TestMessageQueue
 import vinyldns.mysql.TransactionProvider
 import vinyldns.core.TestZoneData.testConnection
-import vinyldns.core.domain.{Fqdn, HighValueDomainError}
+import vinyldns.core.domain.{Encrypted, Fqdn, HighValueDomainError}
```
Comment (JoshSEdwards) [inline modules/api/src/it/scala/vinyldns/api/domain/record/RecordSetServiceIntegrationSpec.scala:36]: Is this needed? It doesn't seem to be used in this file.

**Review by JoshSEdwards — APPROVED** [review]
Looks good to me!

File: modules/api/src/it/scala/vinyldns/api/domain/record/RecordSetServiceIntegrationSpec.scala (hunk 2 above)
Comment (Aravindh-Raju) [reply modules/api/src/it/scala/vinyldns/api/domain/record/RecordSetServiceIntegrationSpec.scala:36]: I haven't completed changes to this file at that point in time. Made the required changes now.

File: modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala (hunk 1 above)
Comment (Aravindh-Raju) [reply modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala:133]: Added tests for `EncryptedSerializer` to verify it's working. Please have a look.

**Review by JoshSEdwards — DISMISSED** [review]

**Review by Jay07GIT — DISMISSED** [review]
Looks good to me.

**Comment by JoshSEdwards** [comment]
Looks good to me. Tested a bunch of things locally using the portal and the API, and didn't see any weird behavior.

**Review by JoshSEdwards — APPROVED** [review]

**Review by arpit4ever — DISMISSED** [review]
Tested locally. Everything looks good.

**Review by arpit4ever — APPROVED** [review]
LGTM

**Pull request merged**